game_patch/misc/g_solid.cpp

```diff
@@ -409,6 +410,53 @@ void g_solid_render_ui()
     }
 }
 
+std::optional<int> bm_get_cache_slot_opt(int handle)
+{
+    int cache_slot = rf::bm_get_cache_slot(handle);
+    return (cache_slot == -1) ? std::nullopt : std::optional<int>{cache_slot};
+}
+
+int bm_get_width(int handle)
+{
+    auto cache_slot = bm_get_cache_slot_opt(handle).value_or(0);
+    if (cache_slot == 0) {
+        return 0;
+    }
+    int index = 27 * cache_slot;
+    return *((uint16_t*)rf::bm_bitmaps + 2 * index + 22);
+}
+
+int bm_get_height(int handle)
+{
+    auto cache_slot = bm_get_cache_slot_opt(handle).value_or(0);
+    if (cache_slot == 0) {
+        return 0;
+    }
+    int index = 27 * cache_slot;
+    return *((uint16_t*)rf::bm_bitmaps + 2 * index + 23);
+}
+
+float bm_calculate_ppm(int handle)
+{
+    const int width = bm_get_width(handle);
+    const int height = bm_get_height(handle);
+
+    if (width <= 0 || height <= 0) {
+        // something weird happened, use default value and avoid divide by zero
+        return 32.0f;
+    }
+
+    const float max_dimension = static_cast<float>(std::max(width, height));
+    return 32.0f * (max_dimension / 256.0f);
+}
+
+using GSolid_SetAutotexture_Type = void __fastcall(rf::GSolid*, float);
+CallHook<GSolid_SetAutotexture_Type> set_geo_crater_ppm_hook{
+    0x00466BD4, [](rf::GSolid* solid, float ppm) {      
+        set_geo_crater_ppm_hook.call_target(solid, bm_calculate_ppm(rf::geomod_crater_texture_handle));
```

Comment:
you always use `level.default_rock_texture` but geomod uses another texture for ice. If someone changes only rock texture it will use wrong ppm for ice. 
Selecting a texture is just a line below. It would be nice to make it work properly for ice too.